pkg/flow/internal/worker/worker_pool.go

```diff
@@ -6,46 +6,52 @@ import (
 	"math/rand"
 	"runtime"
 	"sync"
+	"time"
 )
 
 type Pool interface {
 	// Stop stops the worker pool. It does not wait to drain any internal queues, but it does wait for the currently
 	// running tasks to complete. It must only be called once.
 	Stop()
-	// Submit submits a function to be executed by the worker pool on a random worker. Error is returned if the pool
-	// is unable to accept extra work.
-	Submit(func()) error
-	// SubmitWithKey submits a function to be executed by the worker pool, ensuring that only one
+	// TrySubmit submits a function fn to be executed by the worker pool on a random worker. Error is returned if the
+	// pool is unable to accept extra work within the provided timeout.
+	TrySubmit(fn func(), timeout time.Duration) error
+	// TrySubmitWithKey submits a function to be executed by the worker pool, ensuring that only one
 	// job with given key can be queued at the time. Adding a job with a key that is already queued is a no-op (even if
-	// the submitted function is different). Error is returned if the pool is unable to accept extra work - the caller
-	// can decide how to handle this situation.
-	SubmitWithKey(string, func()) error
+	// the submitted function is different). Error is returned if the pool is unable to accept extra work within the
+	// provided timeout - the caller can decide how to handle this situation.
+	TrySubmitWithKey(key string, fn func(), timeout time.Duration) error
 	// QueueSize returns the number of tasks currently queued.
 	QueueSize() int
+	// DefaultTimeout returns the default timeout that can be used with TrySubmit and TrySubmitWithKey.
+	DefaultTimeout() time.Duration
```

Comment:
I've opted to keeping the timeout configuration as part of the `Pool` so that we don't need to pass a separate timeout field around - and there are quite a few places between controller/loader/modules which would need it.